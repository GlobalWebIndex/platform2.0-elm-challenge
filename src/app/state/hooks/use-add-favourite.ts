import { useMutation, useQueryClient } from '@tanstack/react-query';
import { queryKeyFavourite } from './use-favourite.tsx';
import { addToFavourite } from '../../repository/favourite-helper.ts';

export const useAddFavourite = () => {
  const queryClient = useQueryClient();

  const { mutate: addFavourite, isPending } = useMutation({
    mutationFn: addToFavourite,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: [queryKeyFavourite] });
    },
  });

  return {
    addFavourite,
    isPending,
  };
};
