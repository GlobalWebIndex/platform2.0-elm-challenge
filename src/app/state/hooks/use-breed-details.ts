import { useQuery } from '@tanstack/react-query';
import { fetchBreedDetails } from '../../repository/fetch-breed-details.ts';

export const queryKeyBreedDetails = 'breed-details';
export const useBreedDetails = (id: string) => {
  const { isPending, error, data } = useQuery({
    queryKey: [queryKeyBreedDetails, id],
    queryFn: () => fetchBreedDetails(id),
  });

  return {
    data,
    error,
    isPending,
  };
};
