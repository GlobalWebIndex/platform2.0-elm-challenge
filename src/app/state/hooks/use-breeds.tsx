import { useQuery } from '@tanstack/react-query';
import { fetchBreeds } from '../../repository/fetch-breeds.ts';

export const queryKeyBreeds = 'breeds';
export const useBreeds = () => {
  const { isPending, error, data } = useQuery({
    queryKey: [queryKeyBreeds],
    queryFn: fetchBreeds,
  });

  return {
    data,
    error,
    isPending,
  };
};
