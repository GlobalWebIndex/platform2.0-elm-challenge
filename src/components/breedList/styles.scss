.breeds {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
  gap: 1rem;

  .link {
    padding: 1rem 0;
    border: 1px solid #ccc;
    border-radius: 0.3rem;
    text-align: center;
    text-decoration: none;
    color: #333;

    &:hover {
      background-color: #f9f9f9;
    }
  }
}

@media screen and (min-width: 576px) {
  .breeds {
    gap: 2rem;
  }
}
