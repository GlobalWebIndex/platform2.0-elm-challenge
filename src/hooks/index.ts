import { useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';

export function useData<T>(
  callback: () => Promise<T>,
  handleError?: () => void
) {
  const [data, setData] = useState<T | null>(null);

  useEffect(() => {
    async function fetchData() {
      try {
        const payload = await callback.call(null);

        setData(payload);
      } catch (error) {
        handleError?.();
      }
    }

    if (data === null) fetchData();
    // eslint-disable-next-line
  }, []);

  // Need this specific code structure
  // to achieve the desired typing
  if (data === null) {
    return [false, null] as const;
  }

  return [true, data] as const;
}

export function usePageTitle() {
  const location = useLocation();
  const title = location.pathname.substring(1);
  const defaultTitle = 'My Cat App';

  return title === '' || title.includes('images') ? defaultTitle : title;
}
