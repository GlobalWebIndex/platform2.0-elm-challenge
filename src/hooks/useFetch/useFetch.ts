import {useCallback, useEffect, useMemo, useState} from 'react';

export const useFetch = <T>(
  endpoint: string
): {abort: () => void; data: T | null; error: Error | null; isFetching: boolean} => {
  const [data, setData] = useState<T | null>(null);
  const [isFetching, setIsFetching] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  const abortController = useMemo(() => new AbortController(), []);

  useEffect(() => {
    fetch(`https://api.thecatapi.com/v1/${endpoint}`, {
      signal: abortController.signal
    })
      .then(response => {
        if (!response.ok) {
          throw new Error('Failed to fetch data');
        }

        return response.json() as T;
      })
      .then(response => {
        setData(response);
      })
      .catch((e: unknown) => {
        if ((e as Error).name === 'AbortError') {
          return;
        }

        setError(e as Error);
      })
      .finally(() => {
        setIsFetching(false);
      });

    return (): void => {
      abortController.abort();
    };
  }, [abortController, endpoint]);

  const abort = useCallback<() => void>(() => {
    abortController.abort();
  }, [abortController]);

  return {abort, data, error, isFetching};
};
