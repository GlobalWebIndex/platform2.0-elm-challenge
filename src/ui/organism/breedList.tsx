import { Breed } from "~/routes/router";
import { Link, LinkBox } from "../atoms/link";

type BreedListProps = {
  breeds: Breed[];
};
export function BreedList({ breeds }: BreedListProps) {
  return (
    <ul className="space-y-2 overflow-hidden">
      {breeds.map(({ id, name, origin }) => (
        <li key={id}>
          <LinkBox to={id}>
            <div className="">{name}</div>
            <div className="text-mauve-9 text-sm">{origin}</div>
          </LinkBox>
        </li>
      ))}
    </ul>
  );
}
