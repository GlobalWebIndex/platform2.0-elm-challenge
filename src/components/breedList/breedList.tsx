import React from 'react';
import {Outlet} from 'react-router';
import {Link} from 'react-router-dom';
import type {Breeds} from 'cat-lover/types/breeds';
import {useFetch} from '../../hooks/useFetch';
import {Loader} from '../loader';
import styles from './styles.scss';

export const BreedList = (): JSX.Element => {
  const {data, isFetching} = useFetch<Breeds>('breeds');

  if (isFetching) {
    return <Loader />;
  }

  return (
    <div className={styles.breeds}>
      {(data as Breeds).map(breed => (
        <Link className={styles.link} key={breed.id} to={breed.id}>
          {breed.name}
        </Link>
      ))}
      <Outlet />
    </div>
  );
};
