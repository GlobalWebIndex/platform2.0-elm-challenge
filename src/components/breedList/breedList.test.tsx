import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import {render} from '@testing-library/react';
import type {Breeds} from 'cat-lover/types/breeds';
import * as UseFetch from '../../hooks/useFetch';
import {BreedList} from './breedList';

type UseFetch = typeof UseFetch.useFetch<Breeds>;
type UseFetchModule = Omit<typeof UseFetch, 'useFetch'> & {useFetch: UseFetch};

const mockUseFetch = jest.fn<ReturnType<UseFetch>, Parameters<UseFetch>>();
jest.mock<UseFetchModule>('../../hooks/useFetch', () => ({
  useFetch: (...args) => mockUseFetch(...args)
}));

describe('<BreedList />', () => {
  beforeEach(() => {
    mockUseFetch.mockReturnValue({abort: jest.fn(), data: null, error: null, isFetching: true});
  });

  describe('when data is fetching', () => {
    it('should render Loader', () => {
      const {container} = render(
        <MemoryRouter initialEntries={['/breeds']}>
          <BreedList />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('when data is fetched', () => {
    const breeds: Breeds = [
      {
        country_code: '',
        country_codes: '',
        id: '1',
        life_span: '',
        name: 'Breed 1',
        origin: '',
        temperament: '',
        weight: {
          imperial: '',
          metric: ''
        },
        wikipedia_url: ''
      },
      {
        country_code: '',
        country_codes: '',
        id: '2',
        life_span: '',
        name: 'Breed 2',
        origin: '',
        temperament: '',
        weight: {
          imperial: '',
          metric: ''
        },
        wikipedia_url: ''
      }
    ];

    beforeEach(() => {
      mockUseFetch.mockReturnValue({abort: jest.fn(), data: breeds, error: null, isFetching: false});
    });

    it('should render breeds', () => {
      const {container} = render(
        <MemoryRouter initialEntries={['/breeds']}>
          <BreedList />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });
});
