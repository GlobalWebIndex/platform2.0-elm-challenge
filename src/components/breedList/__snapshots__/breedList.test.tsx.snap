// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BreedList /> when data is fetched should render breeds 1`] = `
<div>
  <div
    class="breeds"
  >
    <a
      class="link"
      href="/1"
    >
      Breed 1
    </a>
    <a
      class="link"
      href="/2"
    >
      Breed 2
    </a>
  </div>
</div>
`;

exports[`<BreedList /> when data is fetching should render Loader 1`] = `
<div>
  <div
    class="loader"
  >
    <svg>
      <use
        href="function SvgIcons(props) {
  return React.createElement(
    'svg', 
    Object.assign({}, props, {'data-file-name': SvgIcons.name})
  );
}#icon-cw"
      />
    </svg>
  </div>
</div>
`;
